### PR TITLE
library: group sets by artist

### PR DESCRIPTION
**Pull request by cafehaine**
This commit groups sets by artist. A quick preview of how it looks in the output:
image

### CODE REVIEW

File: lib/library/html.cc [hunk 1]
```diff
@@ -65,19 +65,38 @@ static void generate_entry(const beatmap_entry &entry, std::ostream &os)
...
 void generate_beatmap_set_listing(const std::vector<beatmap_set> &sets, std::ostream &os)
 {
 	os << head;
 	os << "<link rel=\"stylesheet\" href=\"" << escape{OSHU_WEB_DIRECTORY} << "/style.css\" />";
-	for (const beatmap_set &set : sets)
-		generate_set(set, os);
+	std::string artist = std::string();
+	bool first_set = true;
+	for (const beatmap_set &set : sets) {
+		if (set.artist.compare(artist) == 0)
```
Comment (fmang) [inline lib/library/html.cc]: Or, more simply `if (set.artist == artist)`
It’s not a good condition though, as set.artist could actually be an empty string on the first iteration.

File: lib/library/html.cc [hunk 2]
```diff
@@ -65,19 +65,38 @@ static void generate_entry(const beatmap_entry &entry, std::ostream &os)
...
+			// create a new one
+			if (first_set) {
+				first_set = false;
+				os << "<article>";
+			} else
+				os << "</article><article>";
+
+			os << "<h2>" << escape{set.artist} << "</h2>";
+			artist = set.artist;
+			generate_set(set, os);
```
Comment (fmang) [inline lib/library/html.cc]: In both branches of the if-else you call generate_set, so the call could be moved into the parent block.

File: lib/library/html.cc [hunk 3]
```diff
@@ -65,19 +65,38 @@ static void generate_entry(const beatmap_entry &entry, std::ostream &os)
  */
 static void generate_set(const beatmap_set &set, std::ostream &os)
 {
-	os << "<article>";
-	os << "<h4>" << escape{set.artist} << " - " << escape{set.title} << "</h4><ul>";
+	os << "<h3>" << escape{set.title} << "</h3><ul>";
 	for (const beatmap_entry &entry : set.entries)
 		generate_entry(entry, os);
-	os << "</ul></article>";
+	os << "</ul>";
```
Comment (fmang) [inline lib/library/html.cc]: I feel like we should keep a semantic element to identify a beatmap set, like a `<section>` maybe.

File: lib/library/html.cc [hunk 4]
```diff
@@ -65,19 +65,38 @@ static void generate_entry(const beatmap_entry &entry, std::ostream &os)
...
 }
 
 void generate_beatmap_set_listing(const std::vector<beatmap_set> &sets, std::ostream &os)
 {
 	os << head;
 	os << "<link rel=\"stylesheet\" href=\"" << escape{OSHU_WEB_DIRECTORY} << "/style.css\" />";
-	for (const beatmap_set &set : sets)
-		generate_set(set, os);
+	std::string artist = std::string();
```
Comment (fmang) [inline lib/library/html.cc]: Maybe `previous_artist`?

**Comment by cafehaine** [comment]
I guess you're right, I'll try to work on some CSS before doing any more modifications to the html.

**Pull request closed without merging**